### PR TITLE
419239:Automate Azure DevOps Service Connection Setup with Azure Portal

### PR DESCRIPTION
**Pull request by santosh-gh**
[AB#419239](https://dev.azure.com/defragovuk/93c65135-d16b-49d6-a191-4a5313532779/_workitems/edit/395100)

What this PR does / why we need it:
Automate the creation of ADO service connection using Federated Identity Credentials

Create federated Identity credential in app reg
Create Ado service connection using federated identity credential
Pipeline flag added for creation of the service connection
This is preferred as it does not certificate and no need to update it if it expired
This will replace the existing service connection created using client secret.

Special notes for your reviewer
Any specific actions or notes on review?

Testing
https://dev.azure.com/defragovuk/DEFRA-FFC/_build/results?buildId=632997&view=results

Checklist (please delete before completing or setting auto-complete)
[x ] Story Work items associated (not Tasks)
[x ] Successful testing run(s) link provided
[x ] Title pattern should be {work item number}: {title}
[x ] Description covers all the changes in the PR
This PR contains documentation
This PR contains tests
How does this PR make you feel:

### CODE REVIEW

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json [hunk 1]
```diff
@@ -22,6 +22,12 @@
                     }
                 ]
             },
+            "federartedCredential":{
+                "name": "#{{ tier2ServiceConnectionAppRegName }}-fic",
+                "issuer": "https://vstoken.dev.azure.com/0843dc02-bf94-4c0c-b0ed-bb5f8c829f46",
```
Comment (rajeshkaremane) [inline infra/bootstrap/env/config/app-registrations/tier2-app-registration.json:27]: the organization id should be dynamic based on the env

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json [hunk 2]
```diff
@@ -22,6 +22,12 @@
                     }
                 ]
             },
+            "federartedCredential":{
+                "name": "#{{ tier2ServiceConnectionAppRegName }}-fic",
```
Comment (rajeshkaremane) [inline infra/bootstrap/env/config/app-registrations/tier2-app-registration.json:26]: remove -fic

File: infra/bootstrap/env/config/service-connections/federated-credential-service-connection.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{
+    "data": {
+        "subscriptionId": "#{{ subscriptionId }}",
+        "subscriptionName": "#{{ subscriptionName }}",
+        "environment": "AzureCloud",
+        "scopeLevel": "Subscription"
+    },  
+    "name": "#{{ subscriptionName }}-FSC",
```
[the remark this answers is not in the document]
Comment (santosh-gh) [reply infra/bootstrap/env/config/service-connections/federated-credential-service-connection.json]: removed

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json (hunk 2 above)
Comment (santosh-gh) [reply infra/bootstrap/env/config/app-registrations/tier2-app-registration.json]: removed

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json [hunk 4]
```diff
@@ -22,6 +22,12 @@
...
+                "issuer": "https://vstoken.dev.azure.com/0843dc02-bf94-4c0c-b0ed-bb5f8c829f46",
+                "subject": "sc://defragovuk/DEFRA-FFC/#{{ subscriptionName }}-FSC",
```
[the remark this answers is not in the document]
Comment (santosh-gh) [reply infra/bootstrap/env/config/app-registrations/tier2-app-registration.json]: removed

File: .azuredevops/deploy-platform-bootstrap-env.yaml [hunk 5]
```diff
@@ -154,6 +163,8 @@ extends:
               path: "infra/bootstrap/env/scripts/ado/Initialize-ServiceEndpoint.ps1"
               scriptArguments: > 
                 -ServiceEndpointJsonPath 'infra/bootstrap/env/config/service-connections/tier2-service-connection.json'
+                -FederatedEndpointJsonPath 'infra/bootstrap/env/config/service-connections/federated-credential-service-connection.json'
+                -FederatedCredential $${{parameters.Federated}}
```
Comment (rajeshkaremane) [inline .azuredevops/deploy-platform-bootstrap-env.yaml:167]: why $$?

File: .azuredevops/deploy-platform-bootstrap-env.yaml [hunk 6]
```diff
@@ -141,6 +149,7 @@ extends:
               path: "PowerShellLibrary/Add-AdAppRegistrations.ps1"
               scriptArguments: > 
                 -AppRegJsonPath '$(Pipeline.Workspace)/s/self/infra/bootstrap/env/config/app-registrations/tier2-app-registration.json'
+                -FederatedCredential $${{parameters.Federated}}
```
Comment (rajeshkaremane) [inline .azuredevops/deploy-platform-bootstrap-env.yaml:152]: why $$?

File: .azuredevops/deploy-platform-bootstrap-env.yaml (hunk 5 above)
Comment (santosh-gh) [reply .azuredevops/deploy-platform-bootstrap-env.yaml:167]: discussed, it is required

File: .azuredevops/deploy-platform-bootstrap-env.yaml (hunk 6 above)
Comment (santosh-gh) [reply .azuredevops/deploy-platform-bootstrap-env.yaml:152]: discussed, it is required

**Review by rajeshkaremane — APPROVED** [review]

**Review by rajeshkaremane — APPROVED** [review]

File: scripts/modules/ado/ado.psm1 [hunk 7]
```diff
@@ -278,6 +278,120 @@ Function Set-ServiceEndpoint() {
...
+        
+        Write-Debug "${functionName}:ArmServiceConnection=$($ArmServiceConnection | ConvertTo-Json -Depth 10)"
+        # Create Federated Identity Credential   
+
+        $appReg = Get-AzADApplication -DisplayName $ArmServiceConnection.appRegName   
+
+        $federatedCredentials = Get-AzADAppFederatedCredential -ApplicationObjectId $appReg.id
+        $federatedCredentials | Select-Object -Property Name
+
+        $devopsOrganizationName = $OrgnizationUri.substring(22)
```
Comment (rajeshkaremane) [inline scripts/modules/ado/ado.psm1:332]: can this be renamed? it is confusing with next line

Comment (santosh-gh) [reply scripts/modules/ado/ado.psm1]: renamed

**Review by rajeshkaremane — APPROVED** [review]

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json [hunk 8]
```diff
@@ -1,7 +1,10 @@
 {
     "applications": [
         {
+            "subscriptionName": "#{{ subscriptionName }}",
```
Comment (ajaysoni-cap) [inline infra/bootstrap/env/config/app-registrations/tier2-app-registration.json:4]: Also, do we need these changes?

File: .azuredevops/deploy-platform-bootstrap-env.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (santosh-gh) [reply .azuredevops/deploy-platform-bootstrap-env.yaml]: not required, Removed

File: infra/bootstrap/env/config/app-registrations/tier2-app-registration.json (hunk 8 above)
Comment (santosh-gh) [reply infra/bootstrap/env/config/app-registrations/tier2-app-registration.json]: Not required, cleaned up the code

**Review by rajeshkaremane — DISMISSED** [review]

File: .azuredevops/deploy-platform-bootstrap-env.yaml [hunk 9]
```diff
@@ -15,6 +15,13 @@ parameters:
       - Monitoring - Workspace
       - Policy Assignments
       - Network
+  - name: Federated
+    displayName: "Use Federated Credential"
+    type: string
+    default: "False"
+    values:
```
Comment (ajaysoni-cap) [inline .azuredevops/deploy-platform-bootstrap-env.yaml:22]: change the type for this parameter to be boolean and remove the values please

Comment (santosh-gh) [reply .azuredevops/deploy-platform-bootstrap-env.yaml]: Added Boolean parameter and removed the values

**Review by rajeshkaremane — APPROVED** [review]

**Review by ajaysoni-cap — APPROVED** [review]

**Pull request merged**